stdlib/public/Reflection/TypeRefBuilder.cpp

```diff
@@ -193,36 +193,85 @@ const TypeRef *TypeRefBuilder::lookupSuperclass(const TypeRef *TR) {
   return Unsubstituted->subst(*this, *SubstMap);
 }
 
+static llvm::Optional<StringRef> FindOutermostModuleName(NodePointer Node) {
+  // Breadth first search until we find the module name so we find the outermost
+  // one.
+  llvm::SmallVector<NodePointer, 8> Queue;
+  Queue.push_back(Node);
+  // Instead of removing items from the front of the queue we just iterate over
+  // them.
+  for (size_t i = 0; i < Queue.size(); ++i) {
+    NodePointer Current = Queue[i];
+    if (Current->getKind() == Node::Kind::Module) {
+      if (Current->hasText())
+        return Current->getText();
+      else
+        return {};
+    }
+    for (auto Child : *Current)
+      Queue.push_back(Child);
+  }
+  return {};
+}
+
+void TypeRefBuilder::PopulateFieldTypeInfoCacheWithReflectionInfo(
+    ReflectionInfo &Info) {
+  if (ProcessedReflectionInfos.find(&Info) != ProcessedReflectionInfos.end())
+    return;
+
+  for (auto FD : Info.Field) {
+    if (!FD->hasMangledTypeName())
+      continue;
+    auto CandidateMangledName = readTypeRef(FD, FD->MangledTypeName);
+    if (auto NormalizedName = normalizeReflectionName(CandidateMangledName)) {
+      FieldTypeInfoCache[std::move(*NormalizedName)] = FD;
+    }
+  }
+
+  ProcessedReflectionInfos.emplace(&Info);
+}
+
 RemoteRef<FieldDescriptor>
 TypeRefBuilder::getFieldTypeInfo(const TypeRef *TR) {
   const std::string *MangledName;
-  if (auto N = dyn_cast<NominalTypeRef>(TR))
+  NodePointer Node;
+  Demangler Dem;
+  if (auto N = dyn_cast<NominalTypeRef>(TR)) {
+    Node = N->getDemangling(Dem);
     MangledName = &N->getMangledName();
-  else if (auto BG = dyn_cast<BoundGenericTypeRef>(TR))
+  } else if (auto BG = dyn_cast<BoundGenericTypeRef>(TR)) {
+    Node = BG->getDemangling(Dem);
     MangledName = &BG->getMangledName();
+  }
   else
     return nullptr;
 
   // Try the cache.
   auto Found = FieldTypeInfoCache.find(*MangledName);
   if (Found != FieldTypeInfoCache.end())
     return Found->second;
+  
+  // Heuristic: find the outermost Module node available, and try to parse the 
+  // ReflectionInfos with a matching name first.
+  auto ModuleName = FindOutermostModuleName(Node);
+  // If we couldn't find a module name or the type is imported (__C module) we
+  // don't any useful information on which image to look for the type.
+  if (ModuleName && ModuleName != llvm::StringRef("__C")) {
+    for (auto &Info : ReflectionInfos) {
+      if (ModuleName == Info.Name) {
+        PopulateFieldTypeInfoCacheWithReflectionInfo(Info);
+        Found = FieldTypeInfoCache.find(*MangledName);
+        if (Found != FieldTypeInfoCache.end()) {
+          return Found->second;
+        }
```

Comment:
both calls to `PopulateFieldTypeInfoCacheWithReflectionInfo` are followed by this lookup. Should there be some function, say `FindFieldDescriptor(Info, *MangledName)`, which searches and also handles the population/caching as an implementation detail?